### PR TITLE
Synquacer: refactor hsspi nor driver

### PR DESCRIPTION
**Pull request by masahisak**
This PR fully replaces the existing HSSPI NOR flash driver to newly designed driver.

The new driver implements NOR flash read/program/erase/reset APIs, and separates off the HSSPI controller specific module(hsspi) and NOR flash specific module(nor).
With that, we can add the support of other devices such as NAND flash device.

### CODE REVIEW

File: product/synquacer/module/synquacer_system/src/load_secure_fw.c [hunk 1]
```diff
@@ -69,55 +74,56 @@ typedef struct arm_tf_fip_package_s arm_tf_fip_package_t;
...
-    src = (void *)((uint32_t)fip_package_p +
-     (uint32_t)fip_package_p->fip_toc_entry[BL32_TOC_ENTRY_INDEX].offset_addr);
+    src = CONFIG_SCB_ARM_TF_OFFSET +
+        (uint32_t)fip_package_p->fip_toc_entry[BL32_TOC_ENTRY_INDEX]
+            .offset_addr;
     dst = (void *)SCP_ADDR_TRANS_AREA;
-    memcpy(dst, src, fip_package_p->fip_toc_entry[BL32_TOC_ENTRY_INDEX].size);
+
+    FWK_LOG_INFO(
+        "[FIP] load BL32 src 0x%08lx, dst %08lx, size 0x%lx",
```
Comment (tarek-arm) [inline product/synquacer/module/synquacer_system/src/load_secure_fw.c:95]: This is failing the build the build with the error:
```
scp/product/synquacer/module/synquacer_system/src/load_secure_fw.c:96:9: warning: format specifies type 'unsigned long' but the argument has type 'uint32_t' (aka 'unsigned int') [-Wformat]
        src,
```
Can you please change it to:
```
--- a/product/synquacer/module/synquacer_system/src/load_secure_fw.c
+++ b/product/synquacer/module/synquacer_system/src/load_secure_fw.c
@@ -92,7 +92,7 @@ static void fw_fip_load_bl32(arm_tf_fip_package_t *fip_package_p)
     dst = (void *)SCP_ADDR_TRANS_AREA;
 
     FWK_LOG_INFO(
-        "[FIP] load BL32 src 0x%08lx, dst %08lx, size 0x%lx",
+        "[FIP] load BL32 src 0x%08x, dst %08x, size 0x%x",
         src,
         (uint32_t)dst,
         (uint32_t)fip_package_p->fip_toc_entry[BL32_TOC_ENTRY_INDEX].size);
```

Thanks

**Review by tarek-arm — CHANGES_REQUESTED** [review]

File: product/synquacer/include/qspi_api.h [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+        1, \
+        0, \
+        0, \
+        0, \
+        0, \
+        false, \
+        IO_CODE, \
+        IO_ADDR, \
+        IO_DATA))
+#define COMMAND_TYPE_WRITE(CODE, ADDR, IO_CODE, IO_ADDR, IO_DATA) \
```
Comment (tarek-arm) [inline product/synquacer/include/qspi_api.h:107]: I suggest using QSPI_COMMAND_TYPE_WRITE or QSPI_CMD_TYPE_WRITE so it's more specific to qspi

**Review by tarek-arm — CHANGES_REQUESTED** [review]

File: product/synquacer/module/synquacer_system/src/load_secure_fw.c (hunk 1 above)
Comment (masahisak) [reply product/synquacer/module/synquacer_system/src/load_secure_fw.c:95]: @tarek-arm 
This log is duplicate so I will remove this line.

Comment (masahisak) [reply product/synquacer/module/synquacer_system/src/load_secure_fw.c:95]: @tarek-arm
I found problem of loading bl32, I fixed the process of loading bl32 and I still need this log output.
In my environment, I use gcc-arm-none-eabi-10.3-2021.07 for cross compiler and I encounter build error when I change from "0x%08lx" to "0x%08x". I will use PRIx32/PRIu32 instead in the next version.

File: framework/test/CMakeLists.txt [hunk 3]
```diff
@@ -43,7 +43,7 @@ set(FWK_SCP_ROOT ${FWK_TEST_ROOT}/../..)
 list(APPEND EXTRA_COMPILE_FLAGS -g3)
 list(APPEND EXTRA_COMPILE_FLAGS -Wall)
 list(APPEND EXTRA_COMPILE_FLAGS -Wextra)
-list(APPEND EXTRA_COMPILE_FLAGS -Werror)
+#list(APPEND EXTRA_COMPILE_FLAGS -Werror)
```
[the remark this answers is not in the document]
Comment (masahisak) [reply framework/test/CMakeLists.txt:46]: @tarek-arm
Sorry, this modification is not intended. Undo this change.

File: product/synquacer/include/bootctl_reg.h [hunk 4]
```diff
@@ -0,0 +1,37 @@
+/*
+ * Arm SCP/MCP Software
+ * Copyright (c) 2021-2022, Arm Limited and Contributors. All rights reserved.
```
[the remark this answers is not in the document]
Comment (masahisak) [reply product/synquacer/include/bootctl_reg.h:3]: I will update the year to just 2022 for all newly created files.

File: product/synquacer/include/qspi_api.h (hunk 2 above)
Comment (masahisak) [reply product/synquacer/include/qspi_api.h:107]: I will add QSPI_ prefix.

File: product/synquacer/include/nor/device_nor_mx25.h [hunk 5]
```diff
@@ -0,0 +1,105 @@
+/*
+ * Arm SCP/MCP Software
+ * Copyright (c) 2021-2022, Arm Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
```
[the remark this answers is not in the document]
Comment (masahisak) [reply product/synquacer/include/nor/device_nor_mx25.h:6]: Yes. I will also add the description for other devices.

**Review by tarek-arm — CHANGES_REQUESTED** [review]

File: product/synquacer/scp_ramfw/config_hsspi.c [hunk 6]
```diff
@@ -1,13 +1,45 @@
 /*
  * Arm SCP/MCP Software
- * Copyright (c) 2018-2021, Arm Limited and Contributors. All rights reserved.
+ * Copyright (c) 2021-2022, Arm Limited and Contributors. All rights reserved.
```
[the remark this answers is not in the document]
Comment (masahisak) [reply product/synquacer/scp_ramfw/config_hsspi.c]: @tarek-arm 
Thanks. I also found the wrong copyright year and updated these files.

- product/synquacer/module/hsspi/Module.cmake
- product/synquacer/module/hsspi/src/Makefile

I modified wrong source code comment in product/synquacer/module/synquacer_system/src/synquacer_main.c.

**Pull request merged**